apps/service_providers/llm_service/runnables.py

```diff
@@ -267,25 +272,129 @@ def invoke(
 
         if config.get("configurable", {}).get("save_input_to_history", True):
             chat_message = self.state.save_message_to_history(input, ChatMessageType.HUMAN)
-            if resource_file_ids:
-                chat_message.metadata = chat_message.metadata | resource_file_ids
-                chat_message.save()
+            if human_message_resource_file_ids:
+                self._add_resource_files_to_message_metadata(chat_message, human_message_resource_file_ids)
 
         # Note: if this is not a new chat then the history won't be persisted to the thread
         thread_id = self.state.get_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID)
         if thread_id:
             input_dict["thread_id"] = thread_id
 
         input_dict["instructions"] = self.state.get_assistant_instructions()
-
         response = self._get_response_with_retries(config, input_dict, thread_id)
+
+        output, ai_message_resource_file_ids = self._save_response_annotations(response)
+
         if not thread_id:
             self.state.set_metadata(Chat.MetadataKeys.OPENAI_THREAD_ID, response.thread_id)
 
-        output = response.return_values["output"]
-        self.state.save_message_to_history(output, ChatMessageType.AI)
+        ai_message = self.state.save_message_to_history(output, ChatMessageType.AI)
+        if ai_message_resource_file_ids:
+            self._add_resource_files_to_message_metadata(ai_message, ai_message_resource_file_ids)
```

Comment:
I think this should be part of `save_message_to_history`:
```python
self.state.save_message_to_history(output, ChatMessageType.AI, file_ids)
```

Reply:
Good call